vpr/src/base/vpr_constraints.cpp

```diff
@@ -0,0 +1,45 @@
+#include "vpr_constraints.h"
+
+void VprConstraints::add_constrained_atom(const AtomBlockId blk_id, const PartitionId part_id) {
+    constrained_atoms.insert({blk_id, part_id});
+
+    auto got = constrained_atoms.find(blk_id);
+
+    if (got == constrained_atoms.end()) {
```

Comment:
Each atom can be in only one partition; if it's already in a partition it will be switched to the new partition instead.